src/json.c

```diff
@@ -0,0 +1,1150 @@
+      json_append_member(expr_node, "type", json_mkstring(istr(type_ident(tree_type(t)))));
+      json_append_member(expr_node, "expr", dump_expr(tree_value(t)));
+      break;
+
+   case T_OPEN:
+      json_append_member(expr_node, "cls", json_mkstring("open"));
+      break;
+
+   default:
+      cannot_dump(t, "expr");
+   }
+   return expr_node;
+}
+
+static const char *dump_minify_type(const char *name)
+{
+   static const char *known[] = {
+      "STD.STANDARD.",
+      "IEEE.NUMERIC_STD.",
+      "IEEE.STD_LOGIC_1164.",
+   };
+
+   for (size_t i = 0; i < ARRAY_LEN(known); i++) {
+      const size_t len = strlen(known[i]);
+      if (strncmp(name, known[i], len) == 0) {
+         static char buf[256];
+         checked_sprintf(buf, sizeof(buf), "~%s%%s", name + len);
+         return buf;
+      }
+   }
+
+   return name;
+}
+
+static JsonNode *dump_type(type_t type)
+{
+   JsonNode *type_node = json_mkobject();
+   char *type_name = strdup(type_pp_minify(type, dump_minify_type));
+   if (type_name[0] == '~') // remove first ~ from type if present
+      type_name ++;
+
+   unsigned long type_name_len = strlen(type_name);
+   if (type_name[type_name_len-2] == '%')
+      type_name[type_name_len-2] = '\0';
+
+   if (type_kind(type) == T_SUBTYPE && type_has_ident(type))
+      json_append_member(type_node, "name", json_mkstring(type_name));
+   else if (type_is_array(type) && !type_is_unconstrained(type)) {
+      json_append_member(type_node, "name", json_mkstring(type_name));
+      const int ndims = array_dimension(type);
+      JsonNode *range = json_mkarray();
+      json_append_member(type_node, "range", range);
+      for (int i = 0; i < ndims; i++) {
+         JsonNode *range_obj = json_mkobject();
+         range_t r = range_of(type, i);
+         json_append_member(range_obj, "l", dump_expr(r.left));
+         
+         switch (r.kind) {
+         case RANGE_TO:
+            json_append_member(range_obj, "dir", json_mkstring("to"));
+            json_append_member(range_obj, "r", dump_expr(r.right));
+            break;
+         case RANGE_DOWNTO:
+            json_append_member(range_obj, "dir", json_mkstring("downto"));
+            json_append_member(range_obj, "r", dump_expr(r.right));
+            break;
+         case RANGE_DYN:
+            json_append_member(range_obj, "dir", json_mkstring("dynamic"));
+            json_append_member(range_obj, "r", dump_expr(r.right));
+            break;
+         case RANGE_RDYN:
+            json_append_member(range_obj, "dir", json_mkstring("reverse_dynamic"));
+            json_append_member(range_obj, "r", dump_expr(r.right));
+            break;
+         case RANGE_EXPR:
+            break;
+         }
+         json_append_element(range, range_obj);
+      }
+   }
+   else
+      json_append_member(type_node, "name", json_mkstring(type_name));
+   return type_node;
+}
+
+/*static void dump_op(tree_t t, int indent)
```

Comment:
This is unused too

Reply:
Not sure what it was used for but I removed it. I'll put it back if required